### PR TITLE
Tighten Checkstyle ruleset

### PR DESCRIPTION
**Pull request by sman-81**
- Add rule Indentation (check correct indentation in sources) making rule RegexpSinglelineJava 'Indent of 6 spaces is suspicious' redundant
- Add rule FileTabCharacter (disallow tab chars in sources) making rule RegexpSinglelineJava 'Line contains leading tabs' redundant
- Remove RegexpSinglelineJava rule that checks for JUnit 4 usage "@Test.expected"

Btw, rule RegexpSinglelineJava 'Multiple consecutive space characters' raises false positives in string contents such as `String str = "x  x";`. Kind of hard to fix by way of regex only. There are however currently no cases, so no real issue.

### CODE REVIEW

File: core/src/test/java/org/jdbi/v3/core/h2/TestH2SqlArrays.java [hunk 1]
```diff
@@ -44,7 +44,8 @@ public class TestH2SqlArrays {
         h.useTransaction(th -> {
             th.execute("DROP TABLE IF EXISTS uuids");
             th.execute("CREATE TABLE uuids (u UUID ARRAY)");
-    }));
```
Comment (hgschmie) [inline core/src/test/java/org/jdbi/v3/core/h2/TestH2SqlArrays.java:47]: this one I am meh on. We always had the closing curly and closing brackets on the same line for lambdas.

File: internal/policy/src/main/resources/policy/checkstyle.xml [hunk 2]
```diff
@@ -16,23 +16,25 @@
...
+    <module name="FileTabCharacter"/>
 
     <module name="NewlineAtEndOfFile">
         <property name="lineSeparator" value="lf"/>
     </module>
 
     <module name="RegexpMultiline">
         <property name="format" value="(?s:(\r\n|\r).*)"/>
-        <property name="message" value="Line has CRLF or CR line ending."/>
+        <property name="message" value="Line has CRLF or CR line ending"/>
```
Comment (hgschmie) [inline internal/policy/src/main/resources/policy/checkstyle.xml:29]: not sure about the rewording changes. Why are we removing the '.' ?

File: internal/policy/src/main/resources/policy/checkstyle.xml [hunk 3]
```diff
@@ -70,6 +72,10 @@
             <property name="sortStaticImportsAlphabetically" value="true"/>
         </module>
         <!--<module name="InnerAssignment"/>-->
+        <module name="Indentation">
```
Comment (hgschmie) [inline internal/policy/src/main/resources/policy/checkstyle.xml:75]: isn't that just repeating the defaults? According to https://checkstyle.sourceforge.io/config_misc.html#Indentation, there are six settings. Why are these two special if we don't change the defaults?

File: internal/policy/src/main/resources/policy/checkstyle.xml [hunk 4]
```diff
@@ -123,27 +129,15 @@
             <property name="allowEmptyTypes" value="true"/>
         </module>
 
-        <module name="RegexpSinglelineJava">
```
Comment (hgschmie) [inline internal/policy/src/main/resources/policy/checkstyle.xml:126]: why would we remove the tab check?

File: core/src/test/java/org/jdbi/v3/core/h2/TestH2SqlArrays.java (hunk 1 above)
Comment (sman-81) [reply core/src/test/java/org/jdbi/v3/core/h2/TestH2SqlArrays.java:47]:
```
    public static H2DatabaseExtension h2Extension = H2DatabaseExtension.instance().withInitializer(h ->
        h.useTransaction(th -> {
            th.execute("DROP TABLE IF EXISTS uuids");
            th.execute("CREATE TABLE uuids (u UUID ARRAY)");
        }));
```
Like this?
I changed it.

File: internal/policy/src/main/resources/policy/checkstyle.xml (hunk 2 above)
Comment (sman-81) [reply internal/policy/src/main/resources/policy/checkstyle.xml:29]: I wanted it to be consistent. The style was mixed, some Checkstyle messages with trailing period, some without.

File: internal/policy/src/main/resources/policy/checkstyle.xml (hunk 3 above)
Comment (sman-81) [reply internal/policy/src/main/resources/policy/checkstyle.xml:75]: Yes, these are the defaults. I find some value in spelling defaults out nevertheless, as not everyone will be familiar what they are. But I guess the same could be true in so many other places ... I'll remove them!

File: internal/policy/src/main/resources/policy/checkstyle.xml (hunk 4 above)
Comment (sman-81) [reply internal/policy/src/main/resources/policy/checkstyle.xml:126]: Checkstyle offers a built-in check to achieve this: https://checkstyle.sourceforge.io/config_whitespace.html#FileTabCharacter (added new towards top of `checkstyle.xml`).

It's stricter than the `RegexpSinglelineJava` I removed, as it disallows tab characters alltogether. Note that this won't affect the tab constant `\t`.

**Comment by hgschmie** [comment]
there are now some checkstyle errors in the CI, so something is now checking different than before.

**Comment by sman-81** [comment]
> there are now some checkstyle errors in the CI, so something is now checking different than before.

Hi @hgschmie, these were introduced by your commit "Fix resource leaks in script and sqlobject" (214adffa372cf96af9abb6c46a09652b5669b859) on Nov-4 and my subsequent rebase to master.
The updated Checkstyle configuration looks for incorrect indentation.

File: sqlobject/src/test/java/org/jdbi/v3/sqlobject/customizer/internal/TestTimestamped.java [hunk 5]
```diff
@@ -131,7 +131,7 @@ public void shouldUpdateModifiedTimestamp() {
 
         logNext.set("now");
         personDAO.insert(input);
-        Timestamp insert = insertedSqlTimestamp();
+        final Timestamp insert = insertedSqlTimestamp();
```
Comment (hgschmie) [inline sqlobject/src/test/java/org/jdbi/v3/sqlobject/customizer/internal/TestTimestamped.java:134]: why this change?

File: vavr/src/main/java/org/jdbi/v3/vavr/VavrCollectors.java [hunk 6]
```diff
@@ -24,15 +24,16 @@ private VavrCollectors() {
...
+     * The returned collector will throw {@code IllegalStateException} whenever 2 or more elements
+     * are present in a stream. Null elements are mapped to {@code Option.none()}.
+     *
+     * @param <T> the collected type
+     * @return a {@code Collector} which collects 0 or 1 input elements into an {@code Option<T>}.
+     */
+    public static <T> Collector<T, ?, Option<T>> toOption() {
+        return OptionalCollectors.toOptional(Option::none, Option::of);
+    }
+
```
Comment (hgschmie) [inline vavr/src/main/java/org/jdbi/v3/vavr/VavrCollectors.java:38]: do we now need an empty line before the class closing brace?

File: sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestDocumentation.java [hunk 7]
```diff
@@ -83,9 +83,8 @@ public void testFiveMinuteSqlObjectExample() {
     @Test
     public void testObtainHandleViaOpen() {
         assertThatCode(() -> {
-            try (Handle h = h2Extension.openHandle()) {
-                // nop
-            }
+            Handle h = h2Extension.getJdbi().open();
```
Comment (hgschmie) [inline sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestDocumentation.java:86]: this causes a warning in intellij (Using AutoCloseable Handle without t-w-r...)

File: sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestDefineListParameter.java [hunk 8]
```diff
@@ -58,7 +58,7 @@ public void setUp() {
 
     @Test
     public void testWithBindList() {
-        TestDao testDao = handle.attach(TestDao.class);
+        final TestDao testDao = handle.attach(TestDao.class);
```
Comment (hgschmie) [inline sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestDefineListParameter.java:61]: why this change?

File: core/src/test/java/org/jdbi/v3/core/mapper/RowViewMapperTest.java [hunk 9]
```diff
@@ -34,11 +34,11 @@ public class RowViewMapperTest {
 
     @Test
     public void testRowViewMap() {
-        Handle h = h2Extension.getSharedHandle();
         final Map<Integer, String> expected = new HashMap<>();
         expected.put(1, "SFO");
         expected.put(2, "OAK");
         expected.put(3, "YYZ");
+        Handle h = h2Extension.getSharedHandle();
```
Comment (hgschmie) [inline core/src/test/java/org/jdbi/v3/core/mapper/RowViewMapperTest.java:41]: why did this get moved?

File: core/src/test/java/org/jdbi/v3/core/TestHandle.java [hunk 10]
```diff
@@ -183,7 +183,7 @@ public void testCommitRollback() throws Exception {
     public void testIssue2065() throws Exception {
         h.begin();
         h.execute("insert into something (id, name) values (1, 'Brian')");
-        String value = h.createQuery("select name from something where id = 1").mapToBean(Something.class).one().getName();
+        final String value = h.createQuery("select name from something where id = 1").mapToBean(Something.class).one().getName();
```
Comment (hgschmie) [inline core/src/test/java/org/jdbi/v3/core/TestHandle.java:186]: why final?

**Comment by hgschmie** [comment]
I am not opposed to this; but I want is to be able to "format in intellij" and not fight the code formatter. Is there a risk that this might happen?

File: sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestOnDemandSqlObject.java [hunk 11]
```diff
@@ -134,9 +134,8 @@ public void testIteratorPrepatureClose() throws Exception {
         Spiffy spiffy = jdbi.onDemand(Spiffy.class);
         spiffy.insert(1, "Tom");
 
-        try (ResultIterator<Something> all = spiffy.findAll()) {
```
[the remark this answers is not in the document]
Comment (sman-81) [reply sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestOnDemandSqlObject.java:137]: due to `EmptyBlock` check 😬

File: sqlobject/src/test/java/org/jdbi/v3/sqlobject/customizer/internal/TestTimestamped.java (hunk 5 above)
Comment (sman-81) [reply sqlobject/src/test/java/org/jdbi/v3/sqlobject/customizer/internal/TestTimestamped.java:134]: Due to `VariableDeclarationUsageDistance` check which checks the distance between declaration of variable and its first usage. This check has been in our configuration since 2018.
What strikes me that it only fired now.
We are on the latest version now. Did Checkstyle's behaviour change? There has been some work on this check recently ([release notes](https://checkstyle.sourceforge.io/releasenotes.html)).
I'll increase the `allowedDistance` property from 3 (default) to 10, a fair compromise I think.
The check itself makes sense, having declaration and usage too far apart deteriorates readability.

File: vavr/src/main/java/org/jdbi/v3/vavr/VavrCollectors.java (hunk 6 above)
Comment (sman-81) [reply vavr/src/main/java/org/jdbi/v3/vavr/VavrCollectors.java:38]: no

File: sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestOnDemandSqlObject.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (sman-81) [reply sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestOnDemandSqlObject.java:137]: maybe not a check we want to have. I'll kick it out again.

File: sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestDocumentation.java (hunk 7 above)
Comment (sman-81) [reply sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestDocumentation.java:86]: `EmptyBlock` check is gone and the this change reverted

File: sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestDefineListParameter.java (hunk 8 above)
Comment (sman-81) [reply sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestDefineListParameter.java:61]: `VariableDeclarationUsageDistance` check, same as above

File: core/src/test/java/org/jdbi/v3/core/mapper/RowViewMapperTest.java (hunk 9 above)
Comment (sman-81) [reply core/src/test/java/org/jdbi/v3/core/mapper/RowViewMapperTest.java:41]: `VariableDeclarationUsageDistance` check, same as above

File: core/src/test/java/org/jdbi/v3/core/TestHandle.java (hunk 10 above)
Comment (sman-81) [reply core/src/test/java/org/jdbi/v3/core/TestHandle.java:186]: `VariableDeclarationUsageDistance` again. `final` variables are not checked.

**Comment by sman-81** [comment]
> I am not opposed to this; but I want is to be able to "format in intellij" and not fight the code formatter. Is there a risk that this might happen?

I cannot give a guarantee it won't happen, Henning :) Tighter checks require more discipline. The ruleset we have in place and are fine-tuning with this PR is a sound and balanced compromise IMO, also thanks to your valuable feedback which I really appreciate.

File: core/src/main/java/org/jdbi/v3/core/argument/EnumArgumentFactory.java [hunk 12]
```diff
@@ -73,10 +73,11 @@ private static <A, E extends Enum<E>> Optional<Argument> makeArgument(int nullTy
...
-                return Optional.of(new NullArgument(nullType));
-            }
+        if (value == null) {
+            return Optional.of(new NullArgument(nullType));
+        }
 
-            return config.get(Arguments.class).findFor(attributeType, transform.apply(value));
+        return config.get(Arguments.class).findFor(attributeType, transform.apply(value));
     }
+
```
Comment (hgschmie) [inline core/src/main/java/org/jdbi/v3/core/argument/EnumArgumentFactory.java:82]: are those blank lines now needed?

File: core/src/test/java/org/jdbi/v3/core/TestHandle.java [hunk 13]
```diff
@@ -204,15 +204,12 @@ public void testUseHandleUsesSameHandle() {
...
-        jdbi.useHandle(handle1 ->
-                jdbi.useTransaction(handle2 ->
-                        jdbi.withHandle(handle3 ->
-                                jdbi.inTransaction(handle4 -> {
-                                assertThat(handle1).isSameAs(handle2);
-                                assertThat(handle1).isSameAs(handle3);
-                                assertThat(handle1).isSameAs(handle4);
-                                return null;
-                            }))));
+        jdbi.useHandle(handle1 -> jdbi.useTransaction(handle2 -> jdbi.withHandle(handle3 -> jdbi.inTransaction(handle4 -> {
```
Comment (hgschmie) [inline core/src/test/java/org/jdbi/v3/core/TestHandle.java:207]: I consider the multi-line variant much more readable than the one line variant.

File: internal/policy/src/main/resources/policy/checkstyle.xml [hunk 14]
```diff
@@ -64,8 +70,13 @@
             <property name="sortStaticImportsAlphabetically" value="true"/>
         </module>
         <!--<module name="InnerAssignment"/>-->
+        <module name="Indentation"/>
         <module name="InnerTypeLast"/>
         <module name="InterfaceIsType"/>
+        <module name="JavadocStyle">
+            <property name="checkFirstSentence" value="false"/>
```
Comment (hgschmie) [inline internal/policy/src/main/resources/policy/checkstyle.xml:77]: are those changing the default setting or just showing the default?

File: testing/src/test/java/org/jdbi/v3/testing/junit5/JdbiGenericExtensionTest.java [hunk 15]
```diff
@@ -37,13 +37,13 @@ public class JdbiGenericExtensionTest {
...
-                    }
-                });
+            try (Query query = h.createQuery("SELECT name FROM users ORDER BY id")) {
+                return query.mapTo(String.class).list();
+            }
+        });
 
-        assertThat(userNames).hasSize(2);
-        assertThat(userNames).containsExactly("Alice", "Bob");
+        assertThat(userNames).hasSize(2)
```
Comment (hgschmie) [inline testing/src/test/java/org/jdbi/v3/testing/junit5/JdbiGenericExtensionTest.java:45]: these seem to be unrelated to the checkstyle changes.

File: core/src/main/java/org/jdbi/v3/core/argument/EnumArgumentFactory.java (hunk 12 above)
Comment (sman-81) [reply core/src/main/java/org/jdbi/v3/core/argument/EnumArgumentFactory.java:82]: No :)

File: core/src/test/java/org/jdbi/v3/core/TestHandle.java (hunk 13 above)
Comment (sman-81) [reply core/src/test/java/org/jdbi/v3/core/TestHandle.java:207]: I agree with you. The issue with the original code was indentation width of 8 (should be 4).

File: internal/policy/src/main/resources/policy/checkstyle.xml (hunk 14 above)
Comment (sman-81) [reply internal/policy/src/main/resources/policy/checkstyle.xml:77]: The default is `checkFirstSentence = true` which forces you to end the first Javadoc line with a sequence satisfying `"([.?!][ \t\n\r\f<])|([.?!]$)"`.
Let's say you break a lengthy sentence onto the next line, you'd get a Checkstyle violation.

```
        String commentText = "Henning wrote this";
        Pattern endOfSentenceFormat = Pattern.compile("([.?!][ \t\n\r\f<])|([.?!]$)");
        if (!commentText.isEmpty()
            && !endOfSentenceFormat.matcher(commentText).find()
            && !commentText.startsWith("{@inheritDoc}")) {

            System.err.println("Fail: " + commentText);
        }
```

Comment (sman-81) [reply internal/policy/src/main/resources/policy/checkstyle.xml:77]: I think Checkstyle themselves may have changed the default from `false` to `true`, but I did not dig into it any deeper.

File: testing/src/test/java/org/jdbi/v3/testing/junit5/JdbiGenericExtensionTest.java (hunk 15 above)
Comment (sman-81) [reply testing/src/test/java/org/jdbi/v3/testing/junit5/JdbiGenericExtensionTest.java:45]: Dir entgeht auch nichts oder? 😄
this was flagged as 'code smell' not by Checkstyle but by the Github workflow
